src/amo/components/Addon/index.js

```diff
@@ -90,12 +90,17 @@ export class AddonBase extends React.Component {
     }
   }
 
-  componentWillReceiveProps({ addon: newAddon }) {
-    const { addon: oldAddon, dispatch } = this.props;
+  componentWillReceiveProps({ addon: newAddon, params: newParams }) {
+    const { addon: oldAddon, dispatch, errorHandler, params } = this.props;
+
     const oldAddonType = oldAddon ? oldAddon.type : null;
     if (newAddon && newAddon.type !== oldAddonType) {
       dispatch(setViewContext(newAddon.type));
     }
+
+    if (params.slug !== newParams.slug) {
```

Comment:
I'm not sure if an extension can have the same slug as a theme, but just in case (and to at least render an error rather than not update state) we should check if the `addonType` OR the `slug` doesn't match, because right now `/extensions/foo/` -> `/themes/foo/` wouldn't cause a new dispatch.